### ISSUE TITLE
Crash with ArrayIndexOutOfBoundsException during frame handling

### ISSUE THREAD

**Issue by jsvd** [jruby/jruby#6207]
**Environment Information**
  
Provide at least:
* JRuby version at least 9.2.9.0 up to 9.2.11.1
* linux and macosx

Other relevant info you may wish to add:
* So far I've only replicated this inside [Logstash](https://github.com/elastic/logstash)

**Expected Behavior**

<details>
  <summary>With indy and jit enabled</summary>

```
[2020-05-07T17:17:03,372][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Reload/pipeline_id:test_7, :exception=>"Java::JavaLang::ArrayIndexOutOfBoundsException", :message=>"Index -2 out of bounds for length 20", :backtrace=>[
  "org.jruby.runtime.ThreadContext.popFrame(ThreadContext.java:526)",
  "org.jruby.runtime.ThreadContext.postMethodFrameOnly(ThreadContext.java:965)",
  "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.pipeline_action.reload.RUBY$method$execute$0(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/pipeline_action/reload.rb:43)",
  "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.agent.RUBY$block$converge_state$2(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/agent.rb:343)",
  "org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:138)",
  "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)",
  "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)",
  "org.jruby.runtime.Block.call(Block.java:139)",
  "org.jruby.RubyProc.call(RubyProc.java:318)",
  "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)",
  "java.base/java.lang.Thread.run(Thread.java:834)"
]}
```

</details>

<details>
  <summary>With indy disabled and jit enabled</summary>

```
[2020-05-07T17:23:29,788][ERROR][logstash.agent           ] Failed to execute action {:action=>LogStash::PipelineAction::Reload/pipeline_id:test_8, :exception=>"Java::JavaLang::ArrayIndexOutOfBoundsException", :message=>"Index -2 out of bounds for length 20", :backtrace=>[
  "org.jruby.runtime.ThreadContext.popFrame(ThreadContext.java:526)",
  "org.jruby.runtime.ThreadContext.postMethodFrameOnly(ThreadContext.java:965)",
  "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.pipeline_action.reload.RUBY$method$execute$0(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/pipeline_action/reload.rb:43)",
  "org.jruby.internal.runtime.methods.CompiledIRMethod.call(CompiledIRMethod.java:121)",
  "org.jruby.internal.runtime.methods.MixedModeIRMethod.call(MixedModeIRMethod.java:175)",
  "org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:208)",
  "org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:203)",
  "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.agent.invokeOther4:execute(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/agent.rb:343)",
  "Users.joaoduarte.elastic.logstash.logstash_minus_core.lib.logstash.agent.RUBY$block$converge_state$2(/Users/joaoduarte/elastic/logstash/logstash-core/lib/logstash/agent.rb:343)",
  "org.jruby.runtime.CompiledIRBlockBody.callDirect(CompiledIRBlockBody.java:138)",
  "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)",
  "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)",
  "org.jruby.runtime.Block.call(Block.java:139)",
  "org.jruby.RubyProc.call(RubyProc.java:318)",
  "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)",
  "java.base/java.lang.Thread.run(Thread.java:834)"]}
```

</details>

<details>
  <summary>Without indy and jit.threshold=3000000</summary>

```
[2020-05-07T17:28:45,121][ERROR][logstash.agent           ] An exception happened when converging configuration {:exception=>Java::JavaLang::ArrayIndexOutOfBoundsException, :message=>"Index -1 out of bounds for length 40", :backtrace=>[
 "org.jruby.runtime.ThreadContext.popFrameReal(ThreadContext.java:537)", 
 "org.jruby.runtime.ThreadContext.postYield(ThreadContext.java:1079)", 
 "org.jruby.runtime.IRBlockBody.postYield(IRBlockBody.java:184)", 
 "org.jruby.runtime.MixedModeIRBlockBody.commonYieldPath(MixedModeIRBlockBody.java:139)", 
 "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:60)", 
 "org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)", 
 "org.jruby.runtime.Block.call(Block.java:139)", "org.jruby.RubyProc.call(RubyProc.java:318)", 
 "org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)", 
 "java.base/java.lang.Thread.run(Thread.java:834)"
]}
```

</details>

I haven't yet been able to replicate with `-Djruby.compile.mode=OFF`.
[edit] actually I could: seem the [comment below](https://github.com/jruby/jruby/issues/6207#issuecomment-625381529)

Another stack trace that happens in the same scenario can be seen in the original logstash issue: https://github.com/elastic/logstash/issues/11835

**Comment by jsvd** [comment]
ok it actually got it to crash quite spectacularly with compilation disabled:

<details>
  <summary> Without indy or compilation </summary>

```
2020-05-07 18:04:25,249 Converge PipelineAction::Reload<test_7> ERROR An exception occurred processing Appender plain_console java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 20
	at org.jruby.runtime.ThreadContext.popFrame(ThreadContext.java:526)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.post(InterpretedIRMethod.java:41)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.INTERPRET_METHOD(InterpretedIRMethod.java:125)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.call(InterpretedIRMethod.java:110)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:354)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:143)
	at org.jruby.RubyBasicObject.rbInspect(RubyBasicObject.java:1180)
	at org.jruby.RubyObject.inspect(RubyObject.java:442)
	at org.jruby.RubyHash$4.visit(RubyHash.java:855)
	at org.jruby.RubyHash$4.visit(RubyHash.java:851)
	at org.jruby.RubyHash.visitLimited(RubyHash.java:698)
	at org.jruby.RubyHash.visitAll(RubyHash.java:683)
	at org.jruby.RubyHash.inspectHash(RubyHash.java:846)
	at org.jruby.RubyHash.inspect(RubyHash.java:883)
	at org.jruby.RubyHash.to_s(RubyHash.java:945)
	at org.jruby.RubyHash$INVOKER$i$0$0$to_s.call(RubyHash$INVOKER$i$0$0$to_s.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:141)
	at org.jruby.RubyObject.toRubyString(RubyObject.java:244)
	at org.jruby.RubyObject.toString(RubyObject.java:253)
	at java.base/java.lang.String.valueOf(String.java:2951)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:168)
	at org.logstash.log.StructuredMessage.getFormattedMessage(StructuredMessage.java:83)
	at org.apache.logging.log4j.core.pattern.MessagePatternConverter.format(MessagePatternConverter.java:146)
	at org.apache.logging.log4j.core.pattern.PatternFormatter.format(PatternFormatter.java:38)
	at org.apache.logging.log4j.core.layout.PatternLayout$PatternSerializer.toSerializable(PatternLayout.java:341)
	at org.apache.logging.log4j.core.layout.PatternLayout.toText(PatternLayout.java:240)
	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:225)
	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:59)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.directEncodeEvent(AbstractOutputStreamAppender.java:197)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:190)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:181)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:543)
	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:502)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:485)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:460)
	at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:82)
	at org.apache.logging.log4j.core.Logger.log(Logger.java:162)
	at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2190)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2144)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2127)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2020)
	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1891)
	at org.apache.logging.log4j.spi.AbstractLogger.error(AbstractLogger.java:858)
	at org.logstash.log.LoggerExt.rubyError(LoggerExt.java:123)
	at org.logstash.log.LoggerExt$INVOKER$i$0$1$rubyError.call(LoggerExt$INVOKER$i$0$1$rubyError.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:819)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:208)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:203)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:325)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:160)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:116)
	at org.jruby.runtime.InterpretedIRBlockBody.callDirect(InterpretedIRBlockBody.java:102)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)
	at org.jruby.runtime.Block.call(Block.java:139)
	at org.jruby.RubyProc.call(RubyProc.java:318)
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)
	at java.base/java.lang.Thread.run(Thread.java:834)

2020-05-07 18:04:25,251 Converge PipelineAction::Reload<test_7> ERROR An exception occurred processing Appender plain_rolling java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 20
	at org.jruby.runtime.ThreadContext.popFrame(ThreadContext.java:526)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.post(InterpretedIRMethod.java:41)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.INTERPRET_METHOD(InterpretedIRMethod.java:125)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.call(InterpretedIRMethod.java:110)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:354)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:143)
	at org.jruby.RubyBasicObject.rbInspect(RubyBasicObject.java:1180)
	at org.jruby.RubyObject.inspect(RubyObject.java:442)
	at org.jruby.RubyHash$4.visit(RubyHash.java:855)
	at org.jruby.RubyHash$4.visit(RubyHash.java:851)
	at org.jruby.RubyHash.visitLimited(RubyHash.java:698)
	at org.jruby.RubyHash.visitAll(RubyHash.java:683)
	at org.jruby.RubyHash.inspectHash(RubyHash.java:846)
	at org.jruby.RubyHash.inspect(RubyHash.java:883)
	at org.jruby.RubyHash.to_s(RubyHash.java:945)
	at org.jruby.RubyHash$INVOKER$i$0$0$to_s.call(RubyHash$INVOKER$i$0$0$to_s.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:141)
	at org.jruby.RubyObject.toRubyString(RubyObject.java:244)
	at org.jruby.RubyObject.toString(RubyObject.java:253)
	at java.base/java.lang.String.valueOf(String.java:2951)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:168)
	at org.logstash.log.StructuredMessage.getFormattedMessage(StructuredMessage.java:83)
	at org.apache.logging.log4j.core.pattern.MessagePatternConverter.format(MessagePatternConverter.java:146)
	at org.apache.logging.log4j.core.pattern.PatternFormatter.format(PatternFormatter.java:38)
	at org.apache.logging.log4j.core.layout.PatternLayout$PatternSerializer.toSerializable(PatternLayout.java:341)
	at org.apache.logging.log4j.core.layout.PatternLayout.toText(PatternLayout.java:240)
	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:225)
	at org.apache.logging.log4j.core.layout.PatternLayout.encode(PatternLayout.java:59)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.directEncodeEvent(AbstractOutputStreamAppender.java:197)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.tryAppend(AbstractOutputStreamAppender.java:190)
	at org.apache.logging.log4j.core.appender.AbstractOutputStreamAppender.append(AbstractOutputStreamAppender.java:181)
	at org.apache.logging.log4j.core.appender.RollingFileAppender.append(RollingFileAppender.java:312)
	at org.apache.logging.log4j.core.config.AppenderControl.tryCallAppender(AppenderControl.java:156)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender0(AppenderControl.java:129)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppenderPreventRecursion(AppenderControl.java:120)
	at org.apache.logging.log4j.core.config.AppenderControl.callAppender(AppenderControl.java:84)
	at org.apache.logging.log4j.core.config.LoggerConfig.callAppenders(LoggerConfig.java:543)
	at org.apache.logging.log4j.core.config.LoggerConfig.processLogEvent(LoggerConfig.java:502)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:485)
	at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:460)
	at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:82)
	at org.apache.logging.log4j.core.Logger.log(Logger.java:162)
	at org.apache.logging.log4j.spi.AbstractLogger.tryLogMessage(AbstractLogger.java:2190)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageTrackRecursion(AbstractLogger.java:2144)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessageSafely(AbstractLogger.java:2127)
	at org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:2020)
	at org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:1891)
	at org.apache.logging.log4j.spi.AbstractLogger.error(AbstractLogger.java:858)
	at org.logstash.log.LoggerExt.rubyError(LoggerExt.java:123)
	at org.logstash.log.LoggerExt$INVOKER$i$0$1$rubyError.call(LoggerExt$INVOKER$i$0$1$rubyError.gen)
	at org.jruby.internal.runtime.methods.JavaMethod$JavaMethodN.call(JavaMethod.java:819)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:208)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:203)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:325)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:160)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:116)
	at org.jruby.runtime.InterpretedIRBlockBody.callDirect(InterpretedIRBlockBody.java:102)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)
	at org.jruby.runtime.Block.call(Block.java:139)
	at org.jruby.RubyProc.call(RubyProc.java:318)
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)
	at java.base/java.lang.Thread.run(Thread.java:834)

warning: thread "Converge PipelineAction::Reload<test_7>" terminated with exception (report_on_exception is true):
java.lang.ArrayIndexOutOfBoundsException: Index -2 out of bounds for length 20
	at org.jruby.runtime.ThreadContext.getCurrentScope(ThreadContext.java:300)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:125)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:80)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.INTERPRET_METHOD(InterpretedIRMethod.java:119)
	at org.jruby.internal.runtime.methods.InterpretedIRMethod.call(InterpretedIRMethod.java:110)
	at org.jruby.runtime.callsite.CachingCallSite.cacheAndCall(CachingCallSite.java:354)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:143)
	at org.jruby.RubyObject.toRubyString(RubyObject.java:244)
	at org.jruby.RubyObject.toString(RubyObject.java:253)
	at java.base/java.util.Formatter$FormatSpecifier.printString(Formatter.java:3031)
	at java.base/java.util.Formatter$FormatSpecifier.print(Formatter.java:2908)
	at java.base/java.util.Formatter.format(Formatter.java:2673)
	at java.base/java.util.Formatter.format(Formatter.java:2609)
	at java.base/java.lang.String.format(String.java:2897)
	at org.logstash.execution.ConvergeResultExt$ActionResultExt.create(ConvergeResultExt.java:131)
	at org.logstash.execution.ConvergeResultExt.add(ConvergeResultExt.java:57)
	at org.logstash.execution.ConvergeResultExt$INVOKER$i$2$0$add.call(ConvergeResultExt$INVOKER$i$2$0$add.gen)
	at org.jruby.runtime.callsite.CachingCallSite.call(CachingCallSite.java:203)
	at org.jruby.ir.interpreter.InterpreterEngine.processCall(InterpreterEngine.java:325)
	at org.jruby.ir.interpreter.InterpreterEngine.interpret(InterpreterEngine.java:160)
	at org.jruby.ir.interpreter.Interpreter.INTERPRET_BLOCK(Interpreter.java:116)
	at org.jruby.runtime.InterpretedIRBlockBody.callDirect(InterpretedIRBlockBody.java:102)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:58)
	at org.jruby.runtime.IRBlockBody.call(IRBlockBody.java:52)
	at org.jruby.runtime.Block.call(Block.java:139)
	at org.jruby.RubyProc.call(RubyProc.java:318)
	at org.jruby.internal.runtime.RubyRunnable.run(RubyRunnable.java:105)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

</details>

**Comment by headius** [comment]
This will be tricky to diagnose. The root cause would seem to be that some code is popping a frame when it didn't push one, or someone's popping frames on a thread they don't belong to.

Is there a way to reproduce this?

**Comment by headius** [comment]
@jsvd I also see there are some other traces in the comments to #6184. Could you trim that to the ones that seem relevant and link them here?

**Comment by jsvd** [comment]
hi @headius sorry for the delay in answering, I need to spend some time on this trying to reduce the reproducer, which is still quite flaky and unpredictable, I've gotten very different stack traces depending on the test.
What flags do you recommend me using to facilitate debugging? disabling indy and the compiler? any other debug flag?

**Comment by jsvd** [comment]
I've yet to better understand why, but the problem seems to come from the only place where we start a ruby thread from java, the code is https://github.com/elastic/logstash/blob/master/logstash-core/src/main/java/org/logstash/execution/ShutdownWatcherExt.java#L65-L72

Moving the thread creation from here to ruby itself seems to no longer replicate the problem.

**Comment by jsvd** [comment]
Closing for now as this should be fixed by https://github.com/elastic/logstash/pull/11900, to be released in 7.8.1 onwards.
